File: src/providers/minimax.ts
/**
 * MiniMax provider (OpenAI-compatible Chat Completions API).
 *
 * MiniMax's LLM API (`https://api.minimax.io/v1/chat/completions`) follows
 * the OpenAI wire format, so this provider uses plain `fetch` — no SDK — and
 * implements the same `AIProvider` contract as GroqProvider / NvidiaProvider.
 *
 * The API key is sent as `Authorization: Bearer <key>` (server-side only —
 * the CLI never holds or sends `MINIMAX_API_KEY`).
 *
 * Model: MiniMax-M3 (1M context, tool calling, adaptive thinking). The model
 * id is centralized in src/agents/modelRouter.ts; `reasoning_split: true`
 * keeps thinking blocks out of `content` so tool-call parsing stays clean.
 *
 * Capabilities wired in: chat completion, tool/function calling, usage
 * reporting including cached input tokens (`prompt_tokens_details.cached_tokens`),
 * max output control, cancellation (AbortSignal) and classified, scrubbed
 * `ProviderError`s. Streaming is buffered through `chat` like the other
 * server-side providers (the backend proxy is non-streaming).
 */
import { scrub, ProviderError } from './errors.ts';
import type { AIProvider, ChatMessage, ChatOptions, ChatResult, ModelInfo, StreamEvent, ToolCallParam, ToolDefinition, Usage } from './types.ts';

export const MINIMAX_BASE_URL = 'https://api.minimax.io/v1';

/** Fallback default only — the real default lives in the model router. */
export const DEFAULT_MINIMAX_MODEL = 'MiniMax-M3';

const DEFAULT_TIMEOUT_MS = 60_000;
const DEFAULT_CONTEXT = 1_000_000;

interface MiniMaxResponse {
  choices?: Array<{
    message?: { content?: string | null; tool_calls?: MiniMaxToolCall[] };
    finish_reason?: string;
  }>;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
    prompt_tokens_details?: { cached_tokens?: number };
  };
  error?: { message?: string; type?: string; code?: string };
  /** MiniMax business-level status; non-zero means an error even on HTTP 200. */
  base_resp?: { status_code?: number; status_msg?: string };
}

interface MiniMaxToolCall {
  id?: string;
  type?: string;
  function?: { name?: string; arguments?: string };
}

/** Convert our messages to the OpenAI/MiniMax wire format. */
function toWire(messages: ChatMessage[]): Array<Record<string, unknown>> {
  return messages.map((m) => {
    if (m.role === 'tool') {
      return { role: 'tool', tool_call_id: m.tool_call_id ?? '', content: m.content ?? '' };
    }
    if (m.role === 'assistant' && m.tool_calls && m.tool_calls.length > 0) {
      return {
        role: 'assistant',
        content: m.content ?? null,
        tool_calls: m.tool_calls.map((tc) => ({
          id: tc.id,
          type: 'function',
          function: { name: tc.name, arguments: tc.arguments },
        })),
      };
    }
    return { role: m.role, content: m.content ?? '' };
  });
}

function toUsage(u: MiniMaxResponse['usage']): Usage | undefined {
  if (!u) return undefined;
  const inputTokens = u.prompt_tokens ?? 0;
  const outputTokens = u.completion_tokens ?? 0;
  return {
    inputTokens,
    outputTokens,
    totalTokens: u.total_tokens ?? inputTokens + outputTokens,
    cachedInputTokens: u.prompt_tokens_details?.cached_tokens ?? 0,
  };
}

/** Cached input tokens (MiniMax reports prompt_tokens_details.cached_tokens). */
export function cachedInputTokens(u: MiniMaxResponse['usage']): number {
  return u?.prompt_tokens_details?.cached_tokens ?? 0;
}

/** Classify a failed MiniMax response (HTTP status + body) into a category. */
function classifyFailure(status: number, data: MiniMaxResponse | null): ProviderError['category'] {
  const msg = data?.error?.message ?? data?.base_resp?.status_msg ?? '';
  const code = data?.base_resp?.status_code ?? 0;
  const isQuota = /quota|exhausted|balance|insufficient/i.test(msg);
  const isRate = /rate.?limit|too many|frequency/i.test(msg);
  if (status === 401 || status === 403 || code === 1004) return 'authentication';
  if (status === 429 || isRate) return isQuota ? 'quota_exhausted' : 'rate_limit';
  if (status === 404) return 'unavailable_model';
  if (status === 408) return 'timeout';
  if (status >= 500) return 'server_error';
  if (isQuota) return 'quota_exhausted';
  return 'unknown';
}

export class MiniMaxProvider implements AIProvider {
  readonly id = 'minimax';
  readonly label = 'MiniMax';

  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeoutMs: number;
  private modelId: string;

  constructor(opts: { apiKey: string; model?: string; baseUrl?: string; timeoutMs?: number }) {
    this.apiKey = opts.apiKey;
    this.baseUrl = (opts.baseUrl ?? MINIMAX_BASE_URL).replace(/\/+$/, '');
    this.modelId = opts.model ?? DEFAULT_MINIMAX_MODEL;
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  getModel(): ModelInfo {
    return {
      id: this.modelId,
      contextWindow: DEFAULT_CONTEXT,
      supportedFeatures: ['tool_calls', 'json', 'streaming'],
    };
  }

  setModel(modelId: string): void {
    this.modelId = modelId;
  }

  async listModels(): Promise<string[]> {
    try {
      const res = await fetch(`${this.baseUrl}/models`, {
        headers: { Authorization: `Bearer ${this.apiKey}` },
        signal: AbortSignal.timeout(Math.min(this.timeoutMs, 10_000)),
      });
      if (!res.ok) return [];
      const data = (await res.json()) as { data?: Array<{ id: string }> };
      return (data.data ?? []).map((m) => m.id).sort();
    } catch {
      return [];
    }
  }

  async chat(messages: ChatMessage[], options: ChatOptions = {}): Promise<ChatResult> {
    const body: Record<string, unknown> = {
      model: this.modelId,
      messages: toWire(messages),
      temperature: options.temperature ?? 0.2,
      // reasoning_split keeps <think> blocks out of content (tool-call safe).
      reasoning_split: true,
    };
    // MiniMax recommends max_completion_tokens (max_tokens is deprecated).
    if (options.maxTokens !== undefined) body.max_completion_tokens = options.maxTokens;
    if (options.tools && options.tools.length > 0) body.tools = options.tools.map((t) => t as unknown as Record<string, unknown>);

    const data = await this.post(body, options);
    const choice = data.choices?.[0];
    if (!choice) {
      throw new ProviderError(this.id, 'malformed_response', 'MiniMax returned no completion choices.', 200);
    }
    const msg = choice.message;
    return {
      content: msg?.content ?? null,
      toolCalls: msg?.tool_calls ? fromWireToolCalls(msg.tool_calls) : [],
      usage: toUsage(data.usage),
      finishReason: choice.finish_reason ?? 'stop',
    };
  }

  /** Buffered stream: the backend proxy is non-streaming (single JSON). */
  async *streamChat(messages: ChatMessage[], options: ChatOptions = {}): AsyncIterable<StreamEvent> {
    const result = await this.chat(messages, options);
    if (result.content) yield { type: 'content', content: result.content };
    for (const [index, tc] of result.toolCalls.entries()) {
      yield { type: 'tool_call_delta', index, id: tc.id, name: tc.name, argumentsDelta: tc.arguments };
    }
    yield { type: 'done', usage: result.usage };
  }

  private async post(body: Record<string, unknown>, options: ChatOptions): Promise<MiniMaxResponse> {
    const signals: AbortSignal[] = [];
    if (options.signal) signals.push(options.signal);
    signals.push(AbortSignal.timeout(this.timeoutMs));

    let res: Response;
    try {
      res = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${this.apiKey}` },
        body: JSON.stringify(body),
        signal: AbortSignal.any(signals),
      });
    } catch (err) {
      const isTimeout = err instanceof Error && (err.name === 'TimeoutError' || err.name === 'AbortError');
      if (isTimeout) throw new ProviderError(this.id, 'timeout', 'MiniMax request timed out.');
      throw new ProviderError(this.id, 'network', 'Could not reach the MiniMax API.');
    }

    let data: MiniMaxResponse | null = null;
    try {
      data = JSON.parse(await res.text()) as MiniMaxResponse;
    } catch {
      data = null;
    }

    if (!res.ok) {
      const detail = scrub(data?.error?.message ?? '');
      throw new ProviderError(this.id, classifyFailure(res.status, data), detail || 'MiniMax request failed.', res.status);
    }
    if (!data) {
      throw new ProviderError(this.id, 'malformed_response', 'MiniMax returned an unparseable response.', res.status);
    }
    // MiniMax can return HTTP 200 with a business-level error (base_resp).
    const code = data.base_resp?.status_code ?? 0;
    if (code !== 0) {
      const detail = scrub(data.base_resp?.status_msg ?? `MiniMax error ${code}`);
      throw new ProviderError(this.id, classifyFailure(200, data), detail, 200);
    }
    return data;
  }
}

function fromWireToolCalls(tcs: MiniMaxToolCall[]): ToolCallParam[] {
  return tcs.map((tc) => ({
    id: tc.id ?? `call_${Math.random().toString(36).slice(2, 10)}`,
    name: tc.function?.name ?? 'unknown',
    arguments: tc.function?.arguments ?? '{}',
  }));
}

File: grace/providers/minimax.py
"""MiniMax provider (port of src/providers/minimax.ts) — OpenAI-compatible Chat Completions API."""

import random
import string

import requests

from grace.providers.errors import ProviderError, scrub
from grace.providers.types import ChatMessage, ChatOptions, ChatResult, ModelInfo, StreamEvent, ToolCallParam, Usage, as_chat_messages, as_chat_options

MINIMAX_BASE_URL = "https://api.minimax.io/v1"
DEFAULT_MINIMAX_MODEL = "MiniMax-M3"
DEFAULT_TIMEOUT_MS = 60_000
DEFAULT_CONTEXT = 1_000_000


def _to_wire(messages: list[ChatMessage]) -> list[dict]:
    out = []
    for m in messages:
        if m.role == "tool":
            out.append({"role": "tool", "tool_call_id": m.tool_call_id or "", "content": m.content or ""})
        elif m.role == "assistant" and m.tool_calls:
            out.append({
                "role": "assistant",
                "content": m.content,
                "tool_calls": [
                    {"id": tc.id, "type": "function", "function": {"name": tc.name, "arguments": tc.arguments}}
                    for tc in m.tool_calls
                ],
            })
        else:
            out.append({"role": m.role, "content": m.content or ""})
    return out


def _to_usage(u: dict | None) -> Usage | None:
    if not u:
        return None
    input_tokens = u.get("prompt_tokens") or 0
    output_tokens = u.get("completion_tokens") or 0
    return Usage(
        inputTokens=input_tokens,
        outputTokens=output_tokens,
        totalTokens=u.get("total_tokens") or input_tokens + output_tokens,
        cachedInputTokens=(u.get("prompt_tokens_details") or {}).get("cached_tokens") or 0,
    )


def cached_input_tokens(u: dict | None) -> int:
    return ((u or {}).get("prompt_tokens_details") or {}).get("cached_tokens") or 0


def _random_id() -> str:
    return "call_" + "".join(random.choices(string.ascii_lowercase + string.digits, k=8))


def _from_wire_tool_calls(tcs: list[dict]) -> list[ToolCallParam]:
    return [
        ToolCallParam(
            id=tc.get("id") or _random_id(),
            name=tc.get("function", {}).get("name") or "unknown",
            arguments=tc.get("function", {}).get("arguments") or "{}",
        )
        for tc in tcs
    ]


def _classify_failure(status: int, data: dict | None) -> str:
    err = data or {}
    msg = err.get("error", {}).get("message") or (err.get("base_resp") or {}).get("status_msg") or ""
    code = (err.get("base_resp") or {}).get("status_code") or 0
    is_quota = bool(__import__("re").search(r"quota|exhausted|balance|insufficient", msg, __import__("re").I))
    is_rate = bool(__import__("re").search(r"rate.?limit|too many|frequency", msg, __import__("re").I))
    if status in (401, 403) or code == 1004:
        return "authentication"
    if status == 429 or is_rate:
        return "quota_exhausted" if is_quota else "rate_limit"
    if status == 404:
        return "unavailable_model"
    if status == 408:
        return "timeout"
    if status >= 500:
        return "server_error"
    if is_quota:
        return "quota_exhausted"
    return "unknown"


class MiniMaxProvider:
    id = "minimax"
    label = "MiniMax"

    def __init__(self, api_key: str, model: str | None = None, base_url: str | None = None, timeout_ms: int = DEFAULT_TIMEOUT_MS) -> None:
        self.api_key = api_key
        self.base_url = (base_url or MINIMAX_BASE_URL).rstrip("/")
        self.model_id = model or DEFAULT_MINIMAX_MODEL
        self.timeout_ms = timeout_ms

    def get_model(self) -> ModelInfo:
        return ModelInfo(
            id=self.model_id,
            contextWindow=DEFAULT_CONTEXT,
            supportedFeatures=["tool_calls", "json", "streaming"],
        )

    def set_model(self, model_id: str) -> None:
        self.model_id = model_id

    def list_models(self) -> list[str]:
        try:
            res = requests.get(f"{self.base_url}/models", headers={"Authorization": f"Bearer {self.api_key}"}, timeout=min(self.timeout_ms, 10_000) / 1000)
            if not res.ok:
                return []
            data = res.json()
            return sorted(m.get("id", "") for m in (data.get("data") or []))
        except Exception:
            return []

    def chat(self, messages: list[ChatMessage], options: ChatOptions | None = None) -> ChatResult:
        messages = as_chat_messages(messages)
        options = as_chat_options(options)
        body: dict = {
            "model": self.model_id,
            "messages": _to_wire(messages),
            "temperature": options.temperature if options.temperature is not None else 0.2,
            # reasoning_split keeps <think> blocks out of content (tool-call safe).
            "reasoning_split": True,
        }
        if options.maxTokens is not None:
            body["max_completion_tokens"] = options.maxTokens
        if options.tools:
            body["tools"] = [t.__dict__ for t in options.tools]

        data = self._post(body, options)
        choice = (data.get("choices") or [{}])[0]
        if not choice:
            raise ProviderError(self.id, "malformed_response", "MiniMax returned no completion choices.", 200)
        msg = choice.get("message") or {}
        return ChatResult(
            content=msg.get("content"),
            toolCalls=_from_wire_tool_calls(msg.get("tool_calls") or []),
            usage=_to_usage(data.get("usage")),
            finishReason=choice.get("finish_reason") or "stop",
        )

    def stream_chat(self, messages: list[ChatMessage], options: ChatOptions | None = None):
        # Buffered stream: the backend proxy is non-streaming (single JSON).
        result = self.chat(messages, options)
        if result.content:
            yield StreamEvent(type="content", content=result.content)
        for index, tc in enumerate(result.toolCalls):
            yield StreamEvent(type="tool_call_delta", index=index, id=tc.id, name=tc.name, argumentsDelta=tc.arguments)
        yield StreamEvent(type="done", usage=result.usage)

    def _post(self, body: dict, options: ChatOptions) -> dict:
        try:
            res = requests.post(
                f"{self.base_url}/chat/completions",
                json=body,
                headers={"Content-Type": "application/json", "Authorization": f"Bearer {self.api_key}"},
                timeout=self.timeout_ms / 1000,
            )
        except requests.exceptions.Timeout:
            raise ProviderError(self.id, "timeout", "MiniMax request timed out.")
        except requests.exceptions.RequestException:
            raise ProviderError(self.id, "network", "Could not reach the MiniMax API.")

        data = None
        try:
            data = res.json()
        except Exception:
            data = None

        if not res.ok:
            raw = ((data or {}).get("error") or {}).get("message") or ""
            raise ProviderError(self.id, _classify_failure(res.status_code, data), scrub(raw) or "MiniMax request failed.", res.status_code)
        if data is None:
            raise ProviderError(self.id, "malformed_response", "MiniMax returned an unparseable response.", res.status_code)
        # MiniMax can return HTTP 200 with a business-level error (base_resp).
        code = ((data.get("base_resp") or {}).get("status_code")) or 0
        if code != 0:
            detail = scrub((data.get("base_resp") or {}).get("status_msg") or f"MiniMax error {code}")
            raise ProviderError(self.id, _classify_failure(200, data), detail, 200)
        return data
